### PR TITLE
🌱  Replace exec kubectl with client CreateOrUpdate

### LINKED ISSUE #9696: Replace `kubectl apply` in e2e test framework
In the CAPI e2e test framework currently the `kubectl` binary is called to apply yamls when creating Clusters. This would improve error reporting from this function which may help to debug #9688.

https://github.com/kubernetes-sigs/cluster-api/blob/2caa534c6450670e5b41a8c9d81cfbadac3b0350/test/framework/cluster_proxy.go#L248-L254

The easiest solution is to create objects in the test framework with a `client.Create` call either changing the underlying function. This is the approach in https://github.com/kubernetes-sigs/cluster-api/pull/9077. We can ignore `alreadyExists` errors to make the call retryable.

Alternatively we could use the actual rest call from `kubectl` which has the benefit of actually being an 'apply' call. The only issue with this is needing to import the `k8s.io/kubectl` library and it being [relatively complicated to implement](https://github.com/kubernetes/kubectl/blob/351ba8bb1c50284672b3cd5d78a7e3433b867f6b/pkg/cmd/apply/apply.go#L500).

### PR DESCRIPTION
**Pull request by cahillsf**
**Which issue(s) this PR fixes**:
Fixes #9696

/area ci

### CODE REVIEW

File: test/framework/cluster_proxy.go [hunk 1]
```diff
@@ -247,20 +251,60 @@ func (p *clusterProxy) GetCache(ctx context.Context) cache.Cache {
...
+			Name:      o.GetName(),
+			Namespace: o.GetNamespace(),
+		}
+		existingObject.SetAPIVersion(o.GetAPIVersion())
+		existingObject.SetKind(o.GetKind())
+		labels := labels.Set(o.GetLabels())
+		if labelSelector.Matches(labels) {
+			if err := p.GetClient().Get(ctx, objectKey, existingObject); err != nil {
+				// Expected error -- if the object does not exist, create it
+				if strings.Contains(err.Error(), "not found") {
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply test/framework/cluster_proxy.go:292]: Basically we could do something like 

```go
  if !apierrors.IsNotFound(err) {
    return retErrs = append(retErrs, err)
    continue
  }
  if err := p.GetClient().Create(ctx, &o); err != nil {
    retErrs = append(retErrs, err)
  }
```

File: test/framework/cluster_proxy.go [hunk 2]
```diff
@@ -247,20 +251,60 @@ func (p *clusterProxy) GetCache(ctx context.Context) cache.Cache {
...
-func (p *clusterProxy) Apply(ctx context.Context, resources []byte, args ...string) error {
-	Expect(ctx).NotTo(BeNil(), "ctx is required for Apply")
-	Expect(resources).NotTo(BeNil(), "resources is required for Apply")
-
-	if err := testexec.KubectlApply(ctx, p.kubeconfigPath, resources, args...); err != nil {
-		var exitErr *exec.ExitError
-		if errors.As(err, &exitErr) {
-			return pkgerrors.New(fmt.Sprintf("%s: stderr: %s", err.Error(), exitErr.Stderr))
+// Creates or updates objects using the clusterProxy client.
+func (p *clusterProxy) CreateOrUpdate(ctx context.Context, resources []byte, selector ...string) error {
```
Comment (chrischdi) [inline test/framework/cluster_proxy.go:255]: being able to provide `selector ...string` but only at the end allowing to pass a single one just to make the arg optional looks a bit weird and unclean to me.

We could do it similar to `GetWorkloadCluster` and be able to pass in some `opts ...CreateOrUpdateOption`, which would also make it somewhat extendable if required in future.

File: test/framework/cluster_proxy.go [hunk 3]
```diff
@@ -247,20 +251,60 @@ func (p *clusterProxy) GetCache(ctx context.Context) cache.Cache {
...
+	var retErrs []error
+	objs, err := yaml.ToUnstructured(resources)
+	if err != nil {
 		return err
 	}
 
-	return nil
+	existingObject := &unstructured.Unstructured{}
+	for _, o := range objs {
+		o := o
```
[the remark this answers is not in the document]
Comment (cahillsf) [reply test/framework/cluster_proxy.go:281]: i thought so, but this kept getting flagged in the CI as an issue... struggling to find the results, will remove when i take into account your other feedback and see if it occurs again

File: test/framework/cluster_proxy.go (hunk 2 above)
Comment (cahillsf) [reply test/framework/cluster_proxy.go:255]: agreed, i wasnt too happy with this -- sounds good

**Comment by chrischdi** [comment]
And thanks for tackling this!

File: test/framework/cluster_proxy.go (hunk 2 above)
Comment (cahillsf) [reply test/framework/cluster_proxy.go]: 👋 @chrischdi pushed up some changes here, i think this covers all your feedback.  lmk if the implementation of options follows what you had in mind

**Comment by cahillsf** [comment]
hey @chrischdi 👋 wanted to check back in on this and see if you have any other feedback

**Comment by sbueringer** [comment]
@cahillsf I'll take a look soon. Just slowly working through my PR review backlog. But it's getting shorter

File: test/framework/cluster_proxy.go [hunk 4]
```diff
@@ -89,8 +93,8 @@ type ClusterProxy interface {
 	// GetLogCollector returns the machine log collector for the Kubernetes cluster.
 	GetLogCollector() ClusterLogCollector
 
-	// Apply to apply YAML to the Kubernetes cluster, `kubectl apply`.
-	Apply(ctx context.Context, resources []byte, args ...string) error
+	// Creates or updates objects using the clusterProxy client
```
Comment (chrischdi) [inline test/framework/cluster_proxy.go:96]:
```suggestion
	// CreateOrUpdate creates or updates objects using the clusterProxy client
```

File: test/framework/cluster_proxy.go [hunk 5]
```diff
@@ -247,20 +266,59 @@ func (p *clusterProxy) GetCache(ctx context.Context) cache.Cache {
...
-// Apply wraps `kubectl apply ...` and prints the output so we can see what gets applied to the cluster.
-func (p *clusterProxy) Apply(ctx context.Context, resources []byte, args ...string) error {
-	Expect(ctx).NotTo(BeNil(), "ctx is required for Apply")
-	Expect(resources).NotTo(BeNil(), "resources is required for Apply")
-
-	if err := testexec.KubectlApply(ctx, p.kubeconfigPath, resources, args...); err != nil {
-		var exitErr *exec.ExitError
-		if errors.As(err, &exitErr) {
-			return pkgerrors.New(fmt.Sprintf("%s: stderr: %s", err.Error(), exitErr.Stderr))
+// Creates or updates objects using the clusterProxy client.
```
Comment (chrischdi) [inline test/framework/cluster_proxy.go:269]:
```suggestion
// CreateOrUpdate creates or updates objects using the clusterProxy client.
```

File: test/framework/cluster_proxy.go [hunk 6]
```diff
@@ -103,6 +107,21 @@ type ClusterProxy interface {
 	Dispose(context.Context)
 }
 
+// CreateOrUpdateConfig contains options for use with CreateOrUpdate.
+type CreateOrUpdateConfig struct {
+	LabelSelector string
+}
```
Comment (chrischdi) [inline test/framework/cluster_proxy.go:113]:
```suggestion
// createOrUpdateConfig contains options for use with CreateOrUpdate.
type createOrUpdateConfig struct {
	labelSelector string
}
```

Let's keep this one private.

File: test/framework/cluster_proxy.go [hunk 7]
```diff
@@ -103,6 +107,21 @@ type ClusterProxy interface {
 	Dispose(context.Context)
 }
 
+// CreateOrUpdateConfig contains options for use with CreateOrUpdate.
+type CreateOrUpdateConfig struct {
+	LabelSelector string
```
Comment (chrischdi) [inline test/framework/cluster_proxy.go:112]: I don't want to overengineer this, but can't we directly use a `labels.Selector` here? This way we also get rid of the parsing below.

```suggestion
	labelSelector labels.Selector
```

File: test/framework/cluster_proxy.go [hunk 8]
```diff
@@ -247,20 +266,59 @@ func (p *clusterProxy) GetCache(ctx context.Context) cache.Cache {
...
+	if config.LabelSelector != "" {
+		ls, err := metav1.ParseToLabelSelector(config.LabelSelector)
+		if err != nil {
+			return fmt.Errorf("could not parse selector: %+w", err)
+		}
+		labelSelector, err = metav1.LabelSelectorAsSelector(ls)
+		if err != nil {
+			return fmt.Errorf("could not convert metav1.LabelSelector to labels.Selector: %+w", err)
 		}
+	}
```
Comment (chrischdi) [inline test/framework/cluster_proxy.go:287]:
```suggestion
	if config.labelSelector != nil {
		labelSelector = config.labelSelector
	}
```

With the above changes :-)

File: test/e2e/kcp_adoption.go [hunk 9]
```diff
@@ -138,7 +138,7 @@ func KCPAdoptionSpec(ctx context.Context, inputGetter func() KCPAdoptionSpecInpu
 		Expect(workloadClusterTemplate).ToNot(BeNil(), "Failed to get the cluster template")
 
 		By("Applying the cluster template yaml to the cluster with the 'initial' selector")
-		Expect(input.BootstrapClusterProxy.Apply(ctx, workloadClusterTemplate, "--selector", "kcp-adoption.step1")).ShouldNot(HaveOccurred())
+		Expect(input.BootstrapClusterProxy.CreateOrUpdate(ctx, workloadClusterTemplate, framework.WithLabelSelector("kcp-adoption.step1"))).ShouldNot(HaveOccurred())
```
Comment (chrischdi) [inline test/e2e/kcp_adoption.go:141]: With the other changes (+ moving the must function some more up):

```suggestion
		selector := labels.NewSelector().Add(must(labels.NewRequirement("kcp-adoption.step1", selection.Exists, nil)))
		Expect(input.BootstrapClusterProxy.CreateOrUpdate(ctx, workloadClusterTemplate, framework.WithLabelSelector(selector))).ShouldNot(HaveOccurred())
```

**Comment by chrischdi** [comment]
Sorry for the long delay, I think after this 👍

File: test/framework/cluster_proxy.go (hunk 7 above)
Comment (cahillsf) [reply test/framework/cluster_proxy.go:112]: yep no reason not to 👍

**Comment by cahillsf** [comment]
@chrischdi thanks for the feedback, i think all the comments have been addressed when you have a chance to take another pass

**Review by chrischdi — COMMENTED** [review]
/lgtm

File: test/framework/cluster_proxy.go [hunk 10]
```diff
@@ -247,20 +265,52 @@ func (p *clusterProxy) GetCache(ctx context.Context) cache.Cache {
...
-
-	if err := testexec.KubectlApply(ctx, p.kubeconfigPath, resources, args...); err != nil {
-		var exitErr *exec.ExitError
-		if errors.As(err, &exitErr) {
-			return pkgerrors.New(fmt.Sprintf("%s: stderr: %s", err.Error(), exitErr.Stderr))
-		}
+// CreateOrUpdate creates or updates objects using the clusterProxy client.
+func (p *clusterProxy) CreateOrUpdate(ctx context.Context, resources []byte, opts ...CreateOrUpdateOption) error {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for Create")
+	Expect(resources).NotTo(BeNil(), "resources is required for Create")
```
Comment (sbueringer) [inline test/framework/cluster_proxy.go:271]:
```suggestion
	Expect(ctx).NotTo(BeNil(), "ctx is required for CreateOrUpdate")
	Expect(resources).NotTo(BeNil(), "resources is required for CreateOrUpdate")
```

File: test/framework/cluster_proxy.go [hunk 11]
```diff
@@ -247,20 +265,52 @@ func (p *clusterProxy) GetCache(ctx context.Context) cache.Cache {
...
+	labelSelector := labels.Everything()
+	config := &createOrUpdateConfig{}
+	for _, opt := range opts {
+		opt(config)
+	}
+	if config.labelSelector != nil {
+		labelSelector = config.labelSelector
+	}
+	var retErrs []error
```
Comment (sbueringer) [inline test/framework/cluster_proxy.go:280]: nit: Let's move this down to directly above the for loop

File: test/framework/cluster_proxy.go [hunk 12]
```diff
@@ -247,20 +265,52 @@ func (p *clusterProxy) GetCache(ctx context.Context) cache.Cache {
 	return p.cache
 }
 
-// Apply wraps `kubectl apply ...` and prints the output so we can see what gets applied to the cluster.
-func (p *clusterProxy) Apply(ctx context.Context, resources []byte, args ...string) error {
-	Expect(ctx).NotTo(BeNil(), "ctx is required for Apply")
-	Expect(resources).NotTo(BeNil(), "resources is required for Apply")
-
-	if err := testexec.KubectlApply(ctx, p.kubeconfigPath, resources, args...); err != nil {
```
Comment (sbueringer) [inline test/framework/cluster_proxy.go:255]: Let's drop test/framework/exec/kubectl.go & hack/ensure-kubectl.sh (+ it's usages)

**Review by sbueringer — COMMENTED** [review]
Sorry for the long delay. Just a few smaller findings, otherwise happy to merge

**Pull request merged**